### DISCUSSION TITLE
Required Run Tools not accessible

### DISCUSSION THREAD

**Discussion by Bodekaer** [RageAgainstThePixel/com.openai.unity#287] [Q&A] [answered]
Hi @StephenHodgson , 
I've noticed that if I require specific tools/functions on a Run, which are not defined within Assistant.Tools, then the tool call cannot be found.
Is this intentional, or perhaps it could here default to searching any tool/function in the cache, and not only the specific assistant?

See e.g. the line here, where it fails to find the Run-only tool call:
https://github.com/RageAgainstThePixel/com.openai.unity/blob/8b91c8cb8056bde240d5fd58c9fd7b5dda8cdaa9/OpenAI/Packages/com.openai.unity/Runtime/Assistants/AssistantExtensions.cs#L116

**Comment by StephenHodgson** [comment]
@Bodekaer thanks for reporting. Can you give me a bit more context?

How exactly are you setting this up and how can I reproduce it?

**Comment by Bodekaer** [comment]
@StephenHodgson yes, here is a quick example to easily replicate it:

```
await _thread.CreateMessageAsync("Call TestingCustomRunTool tool");
run = await _thread.CreateRunAsync(new CreateRunRequest(_assistant, 
	tools: new [] { Tool.GetOrCreateTool(this, nameof(TestingCustomRunTool)) }, 
	toolChoice: "required")
);
run = await run.WaitForStatusChangeAsync();
while (run.RequiredAction != null)
{
	var toolOutputs = await _assistant.GetToolOutputsAsync(run.RequiredAction.SubmitToolOutputs.ToolCalls);
	run = await run.SubmitToolOutputsAsync(toolOutputs);
	run = await run.WaitForStatusChangeAsync();
}
```

**Reply by StephenHodgson** [reply]
I believe you have to define all the tools you're going to use with the assistant ahead of time.

I've never known to be able to do it any other way. 
Unless the docs specify otherwise?

**Reply by Bodekaer** [reply]
Hmm. Interesting. I cannot find any documentation that Tools must be defined on Assistant first. 
For Run.Tools it just says `Override the tools the assistant can use for this run. This is useful for modifying the behavior on a per-run basis.`.

The above example with Run-only Tools seems to work fine on the API side at least, in the way that OpenAI tries to call the function, even if I haven't specified it on Assistant.Tools, and only on Run.Tools. 

Also, it's a real use-case, where I don't want the Tool to be available by default on the Assistant, and only on this specific Run should it be able to call the Tool.

I guess, I can then define all tools on the Assistant, and then for all Runs, I'll always have to also include a list of Tools to be called, and exclude the few tools that I don't want to be there by default. 

Let me know if you've seen documentation around the Tool having to be first defined on the Assistant. Can't find any.

**Reply by StephenHodgson** [reply]
Ok, I'll do my best to try to accommodate this in the future

**Reply by Bodekaer** [reply]
Super. Thanks so much. I can see that might be a rather big upgrade. 
I wonder if you can think of a simple temporary quick-fix I can apply, e.g. where I search the `toolCache` instead of `assistant.Tools` in the example code line above?

I only have 1 assistant instance at the moment anyways.

**Reply by Bodekaer** [reply]
Update: It works fine now (without registering Run-only tools in Assistant Tools).

I simply: 
1) Replaced `assistant.Tools` with with `Tool.GetRegisteredTools()` and
2) Added this to `Tool.cs` as well. 
`public static IReadOnlyList<Tool> GetRegisteredTools() => toolCache;`

**Reply by StephenHodgson** [reply]
There should be a more protected way to access to toolcache.

**Reply by Bodekaer** [reply]
Agree. Just temporary solution for now :)

**Reply by StephenHodgson** [reply, marked as answer]
not temporary, there is already the method for this: 

[`Tool.GetAllAvailableTools(...);`](https://github.com/RageAgainstThePixel/com.openai.unity/blob/8b91c8cb8056bde240d5fd58c9fd7b5dda8cdaa9/OpenAI/Packages/com.openai.unity/Runtime/Common/Tool.cs#L225C43-L225C63)

**Reply by StephenHodgson** [reply]
I didn't want people to access the underlying collection directly bc it needs some initialization

**Reply by Bodekaer** [reply]
Ahh yes. I see now. Thanks. It works smoothly using that as well.
